lib/que/leaky_bucket.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module Que
+  # Provide a mechanism to throttle resource usage, by specifying a target budget of
+  # resource time to be consumed over a rolling window.
+  #
+  # This is not thread-safe, and should be used by only one thread at any one time.
+  class LeakyBucket
+    def initialize(window:, budget:)
+      @window = window
+      @budget = budget
+      @ratio = budget / window
+      @remaining = 0.0
+      @last_refill = nil
+    end
+
+    # Observe resource usage from within the provided block. This consumes time that
+    # remains in the bucket, and should only be called after the bucket has been refilled.
+    def observe
+      start = monotonic_now
+      value = yield
+      duration = monotonic_now - start
+      @remaining -= duration
+
+      value
+    end
+
+    # Wait for the bucket to be refilled, given the time that has elapsed since the last
+    # refill. This method will block until the remaining budget has become positive.
+    def refill
+      refill_time = monotonic_now
+      time_since_refill = refill_time - (@last_refill || monotonic_now)
+      grant = @ratio * time_since_refill
+
+      @last_refill = refill_time
+      @remaining = [@budget, @remaining + grant].min
+
+      # Sleep long enough that the subsequent refill would provide a grant large enough to
+      # make our balance positive
+      if @remaining < 0.0
+        sleep(-@remaining / @ratio)
+      end
+    end
+
+    private
+
+    def monotonic_now
+      Process.clock_gettime(Process::CLOCK_MONOTONIC)
+    end
+
+    # Wrap the kernel sleep so we can use rspec to stub this method
+    def sleep(duration)
+      Kernel.sleep(duration)
```

Comment:
This makes me sad. Is it not possible to stub the method even if not defined in this class?

Reply:
The other option here is to promote the `Clock` concept that's used in the tests to be used here too, so define:
```ruby
class MonotonicClock
  def monotonic_now
    Proccess.clock_gettime(Process::CLOCK_MONOTONIC)
  end

  def sleep(duration)
    Kernel.sleep(duration)
  end
end
```
and DI that here:
```ruby
```
class LeakyBucket
  def initialize(..., clock: MonotonicClock.new)
    ...
  end
end
```

Then the stubbing is totally replaced with DI 🤷‍♂

Reply:
Problem is, you don't know if you're stubbing the method some other library might be using. That would be super sad, if for some reason RSpec used sleep (I doubt it would, but still)

Reply:
I did consider doing that clock thing, in fact it was my first thought. But then I thought that was probably my experience in Go talking, rather than ruby.

What's your preference?

Reply:
Apologies for slowness. We discussed elsewhere but my preference is Go-style injection for stubbing, but with the default that it Just Works ™️ if you don't override the thing. Plus putting the clock in the same file since it's not really meant for reuse. So 🚀'd version is 💯